### ISSUE TITLE
Surrogate Keys - fallback

### ISSUE THREAD

**Issue by ostark** [joshangell/Falcon#1]
Hi Josh,

after publishing our [HTTP Caching article](https://blog.fortrabbit.com/mastering-http-caching) some months ago, I had the idea to write a cache plugin for Craft that supports tagging and purging by tag for various Cache proxies or CDNs.

However not every Proxy/CDN does support Surrogate Keys. By storing the relations between the tag and the URL in an internal map this could be emulated. So, instead of purging the tags, we would purge (multiple) URLs - the result should be the same.

The whole thing could also work as local full page cache - without a need for a Proxy/CDN at all - or as an additional cache layer for a CDN, to avoid expensive MISS calls from multiple edge locations.

As my Craft 2 knowledge is very limited and version 3 is my current playground (phpstorm helps a lot to understand the core), I did not consider writing the plugin for 2 and 3.

I'm still in the process of researching and need to find answers to these questions:

1. Which `Events` exist to support a proper on-the-fly-auto-tagging for `Entries`, `Sections`, etc?
2. What are the invalidation quirks?
3. Which "Cache Drivers" should be implemented in the beginning?
4. Which edge cases exist > Do we need a big PURGE ALL button :-)?

---

I would love to hear your option. 

Thanks
Oli

**Comment by joshangell** [comment]
Hi Oli,

I didn’t come across your article in my digging, I’ll have a read - thanks!

I’ve had a few thoughts about whether to support tag purging via a URL map for other systems such as (FastCGI) and came to the tentative conclusion that it wasn’t worth it, at least initially. If we assume that the map can be built dynamically per request, similar to how I am currently tagging things, then we would need to store a number of key/value pairs per request. These could be stored in the native Craft data cache which could then make use of Redis or Memcache etc.

However, this could grow quite large, we’d end up with 22 entries for a simple news index with one `Single`, 10 news `Entries` each with 1 `Asset` and a custom `section:news` tag for new items to purge on ... this doesn’t seem like it will scale great given its similar to how the current template caches implementation works and the issues that has with a lot of content (long running/stuck Tasks).

Equally, on element save we would have to select all URLs by tag from the data store and send `PURGE` requests to each of them. If we have many URLs this can take a while, even in batches. Craft 3 looks promising in this department because we have queues rather than the Craft 2 `Task` implementation which could help alleviate some strain here.

Finally on this front, I was planning that if I did add it it would only do the tag/URL map storage if the driver didn’t support tag-based purging, so as not to slow down the initial backend request.

So - I plan on abstracting my cache driver layers enough that at some point in the future non surrogate key based drives could be added, but I want to get the whole tagging thing right first.

---

On to your bullet points (which I numbered...):

1. In Craft 2 there are some but not enough. We need something to happen each time an element is accessed during render, there isn’t an event in Craft 2 for this due to security concerns, hence my core hack. In Craft 3 I _believe_ we can swap out the core `Template` class for our own and thus replicate [these lines here](https://github.com/craftcms/cms/blob/develop/src/helpers/Template.php#L50-L52) for our own use. [Here is the equivalent place in Craft 2](https://github.com/pixelandtonic/Craft-Release/blob/master/app/etc/templating/BaseTemplate.php#L59-L62).
2. Invalidation wise I was initially planning on using the element save event to trigger a PURGE based on the element ID, then having an interface to allow us to pick other events (e.g. entry on save) and then perform some kind of matching (e.g. is new and in section `news`) to purge a given tag (e.g. `section:news`). The other quirk I considered is things that need to purge in the future, say due to a future `postDate` - initially I planned on hooking into [Scheduler](https://github.com/supercool/Scheduler) to deal with this as that will re-save an entry when its `postDate` is reached.
3. As I said earlier my initial thought was Varnish and Fastly, but @khalwat persuaded me to at least consider FastCGI hence the rambling above about how to effectively support both without hurting the performance of either.
4. Yes I was considering this too ... a big purge all would be helpful, but also I think just an input to purge a given URL - CloudFlare lets you do both and I find that useful.

---

Sorry for the epic reply, I’ve had a lot of thoughts on this front recently!

I would love to collaborate with more people on this if you are interested. @khalwat might be up for building the FastCGI side of things if I can get the abstraction layers into a good place.

Anyway, right now I will be focussing on getting the surrogate key purging method working first for Varnish, then work out the best way of dealing with all of the events Craft fires.

Thanks Oli,
Josh

**Comment by ostark** [comment]
Wow. Thanks for your comprehensive answer!

Obviously, the mapping in front of the php application (on a Proxy, in C + in memory) is more efficient. 

I could provide a driver for [KeyCDN](https://www.keycdn.com/api#purge-zone-tag) and one for a Poorman's Cache that hooks into the Request/Response with a home-brewed mapping.

**Comment by joshangell** [comment]
But still valid for the middle ground use-case :)

–Josh

+44 7590 664 185

angell.io – specialist Craft CMS development and consultancy

**Comment by joshangell** [comment]
Ah, just read your comment fully (it didn't all come through email) - I would be more than happy to have those options covered, thanks! Let's make this a community effort 🙌

**Comment by khalwat** [comment]
Yeah I'd definitely be interested in getting this working with FastCGI Cache too @joshangell 

https://nystudio107.com/blog/static-caching-with-craft-cms
